File: src/client/java/fr/hardel/asset_editor/client/javafx/routes/StudioRouter.java
package fr.hardel.asset_editor.client.javafx.routes;

import fr.hardel.asset_editor.permission.StudioPermissions;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.ReadOnlyObjectProperty;
import javafx.beans.property.SimpleObjectProperty;

import java.util.function.Supplier;

public final class StudioRouter {

    private final ObjectProperty<StudioRoute> currentRoute = new SimpleObjectProperty<>(StudioRoute.NO_PERMISSION);
    private Supplier<StudioPermissions> permissionSupplier = () -> StudioPermissions.NONE;

    public void setPermissionSupplier(Supplier<StudioPermissions> supplier) {
        this.permissionSupplier = supplier;
    }

    public ReadOnlyObjectProperty<StudioRoute> routeProperty() {
        return currentRoute;
    }

    public StudioRoute currentRoute() {
        return currentRoute.get();
    }

    public void navigate(StudioRoute route) {
        if (route == null || route == currentRoute.get()) {
            return;
        }

        if (route == StudioRoute.NO_PERMISSION) {
            currentRoute.set(route);
            return;
        }

        if (permissionSupplier.get().isNone()) {
            redirectNoPermission();
            return;
        }

        if (route == StudioRoute.DEBUG_ITEMS && !permissionSupplier.get().isAdmin()) {
            redirectNoPermission();
            return;
        }

        currentRoute.set(route);
    }

    public void revalidate() {
        var saved = currentRoute.get();
        currentRoute.set(StudioRoute.NO_PERMISSION);
        navigate(saved);
    }

    private void redirectNoPermission() {
        if (currentRoute.get() != StudioRoute.NO_PERMISSION)
            currentRoute.set(StudioRoute.NO_PERMISSION);
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/routes/StudioRouter.kt
package fr.hardel.asset_editor.client.compose.routes

import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import fr.hardel.asset_editor.permission.StudioPermissions

class StudioRouter {

    var currentRoute: StudioRoute by mutableStateOf(StudioRoute.NoPermission)
        private set

    val currentConcept: String by derivedStateOf { currentRoute.concept() }
    val isOverview: Boolean by derivedStateOf { currentRoute.isOverview() }

    var permissionSupplier: () -> StudioPermissions = { StudioPermissions.NONE }

    fun navigate(route: StudioRoute) {
        if (route == currentRoute) return

        if (route == StudioRoute.NoPermission) {
            currentRoute = route
            return
        }

        if (permissionSupplier().isNone) {
            redirectNoPermission()
            return
        }

        if (route == StudioRoute.Debug && !permissionSupplier().isAdmin) {
            redirectNoPermission()
            return
        }

        currentRoute = route
    }

    fun revalidate() {
        val saved = currentRoute
        currentRoute = StudioRoute.NoPermission
        navigate(saved)
    }

    private fun redirectNoPermission() {
        if (currentRoute != StudioRoute.NoPermission)
            currentRoute = StudioRoute.NoPermission
    }
}
